### DISCUSSION TITLE
Parallel measurements/ parallel execution of workers or precedures

### DISCUSSION THREAD

**Discussion by hoderlump** [pymeasure/pymeasure#916] [Q&A] [answered]
Hi,

simple question as this is not immediately clear for me from the documentation:
Is it possible to start measurements in parallel? E.g. I want to do several temperature and voltage measurements at the same time.

The fact that workers are run as threads should make this possible. 
A plus would be to plot them also.

The ManagedWindow example only allows procedures and measurements to be started sequentially. 

Maybe I am missing some obvious solution but i am not sure how many measurements with many different measurement devices in parallel are possible with pymeasure.
This issue deals with the topic but it does not offer a solution: https://github.com/pymeasure/pymeasure/issues/271

Thanks a lot!

**Comment by BenediktBurger** [comment]
As far as I know (I might be wrong), the pymeasure GUI is made to have a single measurement at a time.
A difficulty with simultaneous measurement is, that all these measurements might access the same device at the same time, which will eventually lead to a communication problem as pymeasure does not use locks.

We are working on a different measurement paradigm, where the instruments, the directors (setting values etc.), and the data acquisition programs are independent programs which communicate with each other. That way, you can have several measurements at the same time using a mix of instruments (some of which might be used by several measurements at the same time).
[LECO](https://github.com/pymeasure/leco-protocol) is the name of that communication protocol, we are developing.

I use that system in my lab. A general explanation is in the beginning of #676 (which evolved to define LECO).
For example, I have one data acquisition program (DataLogger) logging every shot of the laser (at 20 Hz), while another one logs every second the mean value of certain parameters (e.g. temperature) to get a long term overview.

An additional side node:
pymeasure publishes the results of an entry of a measurement via a zmq PUB socket. So you could have your main measurement doing everything and then write listeners, which listen to these published data entries and store subsets of that main measurement.

**Reply by bilderbuchi** [reply]
> as pymeasure does not use locks.

...yet; this is WIP, I have a branch going and hope to have a PR ready "soon". See #506

**Comment by hoderlump** [comment]
Thanks for your clear answer here.

You wrote "So you could have your main measurement doing everything." You mean, create one measurement procedure which allocates several measurement channels on several devices? If that is possible, this would be a solution.

**Reply by BenediktBurger** [reply]
What do you mean with "measurement channels"?
My idea was, to have a single procedure, which collects all the information you want.
This procedure publishes (already in pymeasure) via zmq each data point.

Then you can listen to these zmq data points and plot them individually (your measurements).

Maybe you could describe your use case more in detail, such that we could propose a better solution.

**Reply by hoderlump** [reply]
By "measurement channels" I mean, I have several pyvisa devices, e.g. Keithley Multimeters and NI CompactDAQ channels with temperature probes (These are not supported by pymeasure yet unfortunatley but this is no problem here.).

I would open one or more pyvisa ResourceManagers and collect data from all these devices in parallel in one prodecure instance/ thread.

Can I read all that data from that procedure instance/ thread using [`pymeasure.experiment.listeners.Recorder`](https://pymeasure.readthedocs.io/en/latest/api/experiment/listeners.html)?
That would be great.

Btw., I like the LECO project but here I need a stable solution.

**Reply by BenediktBurger** [reply, marked as answer]
ok, now I understand.

Pymeasure supports as many instruments as you like in a single procedure, so it should be no problem.

In the `startup` part of the procedure you connect to all the instruments (via one or more ResourceManagers etc.) and in the `execute` part, you read them as you like.
For example:
~~~python
def startup(self):
    self.device1 = Some_pymeasure_instrument("COM7")
    rm = ResourceManager()
    self.device2 = rm.open_resource("COM9")

def execute(self):
    for i in range(self.iterations):
        data = {
            "val1": self.device1.some_property,
            "val2": self.device2.ask("some command"),
        }
        self.emit('results', data)
        if self.should_stop():
            break
~~~

Then the pymeasure plot window will show you "val1" and "val2".

**Reply by hoderlump** [reply]
excellent, thanks for the project and for your example code here!
